x-pack/solutions/observability/plugins/apm/public/components/app/settings/agent_configurations/agent_configuration_create_edit/settings_page/advanced_config_key_input.tsx

```diff
@@ -23,52 +31,84 @@ export function AdvancedConfigKeyInput({
   removeValidationError: (key: string) => void;
 }) {
   // Handle key inputs with local state to avoid duplicated keys overwriting each other
   const [localKey, setLocalKey] = useState(configKey);
-  const [errorMsg, setErrorMsg] = useState<string | null>(null);
-  const isFormInvalid = localKey !== configKey && Boolean(errorMsg);
+  const [errorType, setErrorType] = useState<ErrorType>(ErrorType.None);
+  const [isFormInvalid, setIsFormInvalid] = useState(false);
 
   useEffect(() => {
     setLocalKey(configKey);
   }, [configKey]);
 
-  const getErrorMsg = (newKey: string) => {
-    if (newKey === '') {
-      return i18n.translate('xpack.apm.agentConfig.settingsPage.keyEmptyError', {
-        defaultMessage: 'Key cannot be empty',
-      });
-    }
-    if (checkIfPredefinedConfigKeyExists(newKey)) {
-      return i18n.translate('xpack.apm.agentConfig.settingsPage.keyPredefinedError', {
-        defaultMessage: 'This key is already predefined in the standard configuration above',
-      });
-    }
-    if (checkIfAdvancedConfigKeyExists(newKey, configKey)) {
-      return i18n.translate('xpack.apm.agentConfig.settingsPage.keyDuplicateError', {
-        defaultMessage: 'This key is already used in another advanced configuration',
-      });
+  const touched = useCallback((key: string) => key !== configKey, [configKey]);
+
+  const getErrorType = useCallback(
+    (key: string) => {
+      if (key === '') {
+        return ErrorType.Empty;
+      }
+      if (checkIfPredefinedConfigKeyExists(key)) {
+        return ErrorType.AlreadyPredefined;
+      }
+      if (touched(key) && checkIfAdvancedConfigKeyExists(key)) {
+        return ErrorType.Duplicated;
+      }
+      return ErrorType.None;
+    },
+    [checkIfAdvancedConfigKeyExists, checkIfPredefinedConfigKeyExists, touched]
+  );
+
+  const errorMsg = useMemo(() => {
+    switch (errorType) {
+      case ErrorType.Empty:
+        return i18n.translate('xpack.apm.agentConfig.settingsPage.keyEmptyError', {
+          defaultMessage: 'Key cannot be empty',
+        });
+      case ErrorType.AlreadyPredefined:
+        return i18n.translate('xpack.apm.agentConfig.settingsPage.keyPredefinedError', {
+          defaultMessage: 'This key is already predefined in the standard configuration above',
+        });
+      case ErrorType.Duplicated:
+        return i18n.translate('xpack.apm.agentConfig.settingsPage.keyDuplicateError', {
+          defaultMessage: 'This key is already used in another advanced configuration',
+        });
+      default:
+        return '';
     }
-    return null;
-  };
+  }, [errorType]);
 
-  const handleKeyChange = (newKey: string) => {
-    const errorKey = `key${id}`;
-    const formErrorMsg = getErrorMsg(newKey);
+  useEffect(() => {
+    const errorId = `key${id}`;
+    const isTouched = touched(localKey);
+    const newErrorType = getErrorType(localKey);
+    const hasValidationErrors = newErrorType !== ErrorType.None;
```

Comment:
I don't understand why you needed to add this `errorType` state. Isn't it exactly that same we had we the `errorMsg`?  The logic got a bit too complicated because of this?

```suggestion
    const formErrorMsg = getErrorMsg(newKey);
    const hasValidationErrors = formaErrorMsg !== errorMsg;
```

Reply:
I separated logic while trying different approaches to validate on submit but ended up not using it. However, I decided to keep it separated since it seemed cleaner to me. 

In the end, I see that it adds unnecessary boilerplate to the code. Thanks for pointing it out. Removed https://github.com/elastic/kibana/pull/232206/commits/69d60e4b96742e3e9eb40526196e534f3c443fd7